pkg/operator/controllers/imageconfig/image_controller.go

```diff
@@ -0,0 +1,149 @@
+		// controller is disabled
+		return reconcile.Result{}, nil
+	}
+
+	// Check for cloud type
+	requiredRegistries, err := getCloudAwareRegistries(instance)
+	if err != nil {
+		// Not returning error as it will requeue again
+		return reconcile.Result{}, nil
+	}
+
+	// Get image.config yaml
+	imageconfig, err := r.configcli.ConfigV1().Images().Get(ctx, request.Name, metav1.GetOptions{})
+	if err != nil {
+		return reconcile.Result{}, err
+	}
+
+	// Fail fast if both are not nil
+	if imageconfig.Spec.RegistrySources.AllowedRegistries != nil && imageconfig.Spec.RegistrySources.BlockedRegistries != nil {
+		err := errors.New("both AllowedRegistries and BlockedRegistries are present")
+		return reconcile.Result{}, err
+	}
+
+	removeDuplicateRegistries := func(item string) bool {
+		for _, v := range requiredRegistries {
+			if strings.EqualFold(item, v) {
+				return false
+			}
+		}
+		return true
+	}
+
+	// Append to allowed registries
+	if imageconfig.Spec.RegistrySources.AllowedRegistries != nil {
+
+		imageconfig.Spec.RegistrySources.AllowedRegistries = filterSliceInPlace(imageconfig.Spec.RegistrySources.AllowedRegistries, removeDuplicateRegistries)
+		imageconfig.Spec.RegistrySources.AllowedRegistries = append(imageconfig.Spec.RegistrySources.AllowedRegistries, requiredRegistries...)
+	}
+
+	// Remove from blocked registries
+	if imageconfig.Spec.RegistrySources.BlockedRegistries != nil {
+		imageconfig.Spec.RegistrySources.BlockedRegistries = filterSliceInPlace(imageconfig.Spec.RegistrySources.BlockedRegistries, removeDuplicateRegistries)
+	}
+
+	// Update image config registry
+	_, err = r.configcli.ConfigV1().Images().Update(ctx, imageconfig, metav1.UpdateOptions{})
+	return reconcile.Result{}, err
+}
+
+// SetupWithManager setup the manager
+func (r *Reconciler) SetupWithManager(mgr ctrl.Manager) error {
+	imagePredicate := predicate.NewPredicateFuncs(func(o client.Object) bool {
+		return o.GetName() == imageConfigResource
+	})
+
+	return ctrl.NewControllerManagedBy(mgr).
+		For(&configv1.Image{}, builder.WithPredicates(imagePredicate)).
+		Named(controllers.ImageConfigControllerName).
+		Complete(r)
+}
+
+// Switch case to ensure the correct registries are added depending on the cloud environment (Gov or Public cloud)
+func getCloudAwareRegistries(instance *arov1alpha1.Cluster) ([]string, error) {
+	var requiredRegistries []string
+	switch instance.Spec.AZEnvironment {
+	case azureclient.PublicCloud.Environment.Name:
+		requiredRegistries = []string{instance.Spec.ACRDomain, "arosvc." + instance.Spec.Location + ".data." + azure.PublicCloud.ContainerRegistryDNSSuffix}
+
+	case azureclient.USGovernmentCloud.Environment.Name:
+		requiredRegistries = []string{instance.Spec.ACRDomain, "arosvc." + instance.Spec.Location + ".data." + azure.USGovernmentCloud.ContainerRegistryDNSSuffix}
+
+	default:
+		err := errors.New("cloud environment %s is not supported" + instance.Spec.AZEnvironment)
+		return nil, err
```

Comment:
```suggestion
		return nil, fmt.Errorf("cloud environment %s is not supported", instance.Spec.AZEnvironment)
```

There is "AZEnvironment is Unset" test case which gives false sense of coverage, but does not test the error.

Another reason to have a separate test for `getCloudAwareRegistries` as suggested before.